cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/RunConstraintsChecker.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright © 2015 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.internal.app.runtime.schedule;
+
+import co.cask.cdap.api.schedule.Schedule;
+import co.cask.cdap.app.store.Store;
+import co.cask.cdap.internal.app.runtime.ProgramOptionConstants;
+import co.cask.cdap.internal.app.store.RunRecordMeta;
+import co.cask.cdap.proto.Id;
+import co.cask.cdap.proto.ProgramRunStatus;
+import com.google.common.base.Predicate;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Checks whether all schedule run constraints are met for the schedule to run.
+ */
+class RunConstraintsChecker {
+  private static final Logger LOG = LoggerFactory.getLogger(RunConstraintsChecker.class);
+
+  private final Store store;
+
+  RunConstraintsChecker(Store store) {
+    this.store = store;
+  }
+
+  /**
+   * Checks if the run requirements for the specified schedule and program are all satisfied.
+   *
+   * @param programId the id of the program to check
+   * @param schedule the schedule to check
+   * @return whether all run requirements are satisfied
+   */
+  public boolean checkSatisfied(Id.Program programId, Schedule schedule) {
+    String scheduleName = schedule.getName();
+    Integer maxRuns = schedule.getRunConstraints().getMaxConcurrentRuns();
+    Predicate<RunRecordMeta> scheduleFilter = getScheduleFilter(scheduleName);
+    if (maxRuns != null) {
+      try {
+        int numActive = 0;
+        List<RunRecordMeta> runs = store.getRuns(programId, ProgramRunStatus.RUNNING, 0, Long.MAX_VALUE, maxRuns,
+                                                 scheduleFilter);
+        numActive += runs.size();
+        runs = store.getRuns(programId, ProgramRunStatus.SUSPENDED, 0, Long.MAX_VALUE, maxRuns,
```

Comment:
the thought was that somebody probably suspends a program because they want to limit resource usage, or some other system needs to be fixed, or some other reason that makes it not a good time to run the program. So it's probably not desired for another run to kick off. Do you think we shouldn't count suspended programs?